app/services/totp_creator.go

```diff
@@ -0,0 +1,48 @@
+package services
+
+import (
+	"bytes"
+	"image/png"
+
+	"github.com/keratin/authn-server/app/data"
+	"github.com/keratin/authn-server/lib/route"
+	"github.com/pkg/errors"
+	"github.com/pquerna/otp/totp"
+)
+
+// func TOTPCreator(
+// 	accountStore data.AccountStore, refreshTokenStore data.RefreshTokenStore, keyStore data.KeyStore, actives data.Actives, cfg *app.Config, reporter ops.ErrorReporter,
+// 	accountID int, audience *route.Domain, existingToken *models.RefreshToken,
+// ) (string, string, error) {
+
+//TOTPCreator handles the creation and storage of new TOTP tokens
+func TOTPCreator(accountStore data.AccountStore, totpCache data.TOTPCache, accountID int, audience *route.Domain) (*bytes.Buffer, error) {
+
+	account, err := AccountGetter(accountStore, accountID)
+	if err != nil {
+		return nil, err
+	}
+
+	//Generate totp key
+	key, err := totp.Generate(totp.GenerateOpts{
+		Issuer:      audience.Hostname,
+		AccountName: account.Username,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if err := totpCache.CacheTOTPSecret(account.ID, []byte(key.Secret())); err != nil {
+		return nil, errors.Wrap(err, "TOTPCreator")
+	}
+
+	//Convert to png
+	var buf bytes.Buffer
+	img, err := key.Image(200, 200)
+	if err != nil {
+		return nil, err
+	}
+	png.Encode(&buf, img)
```

Comment:
> ah, yep, that's what i'm referring to. the value being encoded with QR [happens to be a URI](https://github.com/pquerna/otp/blob/master/totp/totp.go#L198) containing information like the secret.

Reply:
3-4kb seems acceptable (and surprising) 👍 

> Or perhaps we could default to a certain size but accept a "size" parameter from the http request to return the PNG rendered to a certain size maybe?

Yeah that'd be easy to retrofit later. Let's wait for the need before implementing custom sizes.